modules/core/include/opencv2/core/types.hpp

```diff
@@ -1895,13 +1895,39 @@ Rect_<_Tp>& operator -= ( Rect_<_Tp>& a, const Size_<_Tp>& b )
 template<typename _Tp> static inline
 Rect_<_Tp>& operator &= ( Rect_<_Tp>& a, const Rect_<_Tp>& b )
 {
-    _Tp x1 = std::max(a.x, b.x);
-    _Tp y1 = std::max(a.y, b.y);
-    a.width = std::min(a.x + a.width, b.x + b.width) - x1;
-    a.height = std::min(a.y + a.height, b.y + b.height) - y1;
+    if (a.empty() || b.empty()) {
+        a = Rect();
+        return a;
+    }
+    _Tp x1, y1, x1_min, y1_min;
+    if (a.x < b.x) {
+        x1_min = a.x;
+        x1 = b.x;
+    } else {
+        x1_min = b.x;
+        x1 = a.x;
+    }
+    if (a.y < b.y) {
+        y1_min = a.y;
+        y1 = b.y;
+    } else {
+        y1_min = b.y;
+        y1 = a.y;
+    }
+    // Maximum possible width and height.
+    const _Tp width_max = std::min(a.width, b.width);
+    const _Tp height_max = std::min(a.height, b.height);
+    // Deal with cases that could overflow (x1 - x1_min) or (y1 - y1_min).
+    if ((x1_min < 0 && x1_min + width_max < x1) ||
+        (y1_min < 0 && y1_min + height_max < y1)) {
+        a = Rect();
+        return a;
```

Comment:
1. This is for integers only. I believe we don't want to change existed logic for floating point computation.

2. Why is overflow case return empty `Rect()`? What is the logic behind that?
IMHO, we should raise an error/exception if we detect overflow/underflow in our operations.
(however we can adapt `&` operator, but it would be hard to adapt others operations, e.g `|` below).

Reply:
1. This is also for floats and doubles: I added a test for it.
2. I added way more explanations, sorry for the confusion. The goal is to avoid overflow computations.

I agree it is impossible to add to other functions (except to .contains). It would solve more cases for operator|=. But we indeed cannot solve cv::Rect(-INT_MAX,-INT_MAX,INT_MAX,INT_MAX) | cv::Rect(INT_MAX,INT_MAX,INT_MAX,INT_MAX)

The goal of this pull request is to solve drawing issues when there is a huge rectangle: you first crop it to right around your image, and use that cropped version.